### PR TITLE
Proxy support for connecting to twitter

### LINKED ISSUE #7: Add proxy support
Twitter river supports proxy setting. See https://github.com/elasticsearch/elasticsearch-river-twitter#using-a-proxy

Would be nice to have it in Twitter Input as well.

### PR DESCRIPTION
**Pull request by purbon**
It Fixes #7 by patching the last version of the twitter gem to work properly with proxies. As soon as there is a new twitter gem released, proxy support should be fix, just opened an issue upstream https://github.com/sferik/twitter/issues/741 to gather feedback to wrap up everything.

### CODE REVIEW

File: lib/logstash/inputs/twitter.rb [hunk 1]
```diff
@@ -95,11 +100,11 @@ def run(queue)
         end
       end # client.filter
     rescue Twitter::Error::TooManyRequests => e
-      @logger.warn("Twitter too many requests error, sleeping for #{e.rate_limit.reset_in}s")
+      puts("Twitter too many requests error, sleeping for #{e.rate_limit.reset_in}s")
```
Comment (suyograo) [inline lib/logstash/inputs/twitter.rb]: why change this to `puts`?

File: lib/logstash/inputs/twitter.rb [hunk 2]
```diff
@@ -95,11 +100,11 @@ def run(queue)
...
       Stud.stoppable_sleep(e.rate_limit.reset_in) { stop? }
       retry
     rescue => e
-      @logger.warn("Twitter client error", :message => e.message, :exception => e, :backtrace => e.backtrace)
+      puts("Twitter client error", :message => e.message, :exception => e, :backtrace => e.backtrace)
```
Comment (suyograo) [inline lib/logstash/inputs/twitter.rb]: same here `puts`?

File: lib/logstash/inputs/twitter.rb (hunk 1 above)
Comment (purbon) [inline lib/logstash/inputs/twitter.rb]: Obviously a debug leftover, thanks for noticing. Will fix.

File: lib/logstash/inputs/twitter.rb (hunk 2 above)
Comment (purbon) [inline lib/logstash/inputs/twitter.rb]: Same

**Comment by purbon** [comment]
Testing this changes, I ecounter this error:

```
vagrant@vagrant-ubuntu-trusty-64:~$ sudo tail /var/log/squid3/access.log 
1447074498.341      0 10.0.2.2 NONE/501 4001 POST https://stream.twitter.com/1.1/statuses/filter.json? - HIER_NONE/- text/html
1447074498.350      0 10.0.2.2 NONE/501 4001 POST https://stream.twitter.com/1.1/statuses/filter.json? - HIER_NONE/- text/html
1447074498.358      0 10.0.2.2 NONE/501 4001 POST https://stream.twitter.com/1.1/statuses/filter.json? - HIER_NONE/- text/html
1447074498.367      0 10.0.2.2 NONE/501 4005 POST https://stream.twitter.com/1.1/statuses/filter.json? - HIER_NONE/- text/html
1447074498.374      0 10.0.2.2 NONE/501 4009 POST https://stream.twitter.com/1.1/statuses/filter.json? - HIER_NONE/- text/html
1447074498.384      0 10.0.2.2 NONE/501 4009 POST https://stream.twitter.com/1.1/statuses/filter.json? - HIER_NONE/- text/html
1447074498.392      0 10.0.2.2 NONE/501 4009 POST https://stream.twitter.com/1.1/statuses/filter.json? - HIER_NONE/- text/html
1447074498.399      0 10.0.2.2 NONE/501 4005 POST https://stream.twitter.com/1.1/statuses/filter.json? - HIER_NONE/- text/html
1447074498.408      0 10.0.2.2 NONE/501 4001 POST https://stream.twitter.com/1.1/statuses/filter.json? - HIER_NONE/- text/html
1447074498.420      0 10.0.2.2 NONE/501 4001 POST https://stream.twitter.com/1.1/statuses/filter.json? - HIER_NONE/- text/html
```

when willing to proxy through an squid3 server. This means the connection is made to the proxy, but failing in the process of reaching the content to the internet. Any idea what might be wrong in the squid config?

**Comment by purbon** [comment]
When testing with polipo I get this errors on the proxy side.

```
Nov  9 13:45:44 vagrant-ubuntu-trusty-64 polipo[1717]: Couldn't parse client's request line
Nov  9 13:50:36 vagrant-ubuntu-trusty-64 polipo[1717]: Server returned weak ETag -- ignored.
Nov  9 13:50:42 vagrant-ubuntu-trusty-64 polipo[1717]: Couldn't parse ETag.
Nov  9 13:50:43 vagrant-ubuntu-trusty-64 polipo[1717]: Couldn't parse ETag.
Nov  9 13:50:46 vagrant-ubuntu-trusty-64 polipo[1717]: Unsupported Cache-Control directive s-max-age -- ignored.
Nov  9 13:51:01 vagrant-ubuntu-trusty-64 polipo[1717]: Unsupported Cache-Control directive stale-while-revalidate -- ignored.
Nov  9 13:51:03 vagrant-ubuntu-trusty-64 polipo[1717]: Unsupported Cache-Control directive stale-while-revalidate -- ignored.
Nov  9 13:51:03 vagrant-ubuntu-trusty-64 polipo[1717]: Unsupported Cache-Control directive stale-if-error -- ignored.
Nov  9 13:51:06 vagrant-ubuntu-trusty-64 polipo[1717]: Unsupported Cache-Control directive stale-while-revalidate -- ignored.
Nov  9 13:51:06 vagrant-ubuntu-trusty-64 polipo[1717]: message repeated 3 times: [ Unsupported Cache-Control directive stale-while-revalidate -- ignored.]
Nov  9 13:51:07 vagrant-ubuntu-trusty-64 polipo[1717]: Unsupported Cache-Control directive s-max-age -- ignored.
```

does this means the proxy string is buggy? or is the proxy configuration?

**Comment by purbon** [comment]
works good, but needs tricky configurations that might be proxy dependant to forward the stuff to the stream api. mostly related to chunck size for http streaming.

**Comment by purbon** [comment]
@untergeek @suyograo would you guys be so nice to review this?

**Comment by untergeek** [comment]
@purbon I have it "running," but all I see is a lot of messages like this:

```ruby
{:timestamp=>"2015-11-17T18:03:46.165000-0700", :message=>"Unrecognized SSL message, plaintext connection?", :exception=>#<OpenSSL::SSL::SSLError: Unrecognized SSL message, plaintext connection?>, :backtrace=>["org/jruby/ext/openssl/SSLSocket.java:210:in `connect'", "/Users/buh/WORK/logstash-plugins/logstash-input-twitter/lib/logstash/inputs/twitter/patches.rb:30:in `stream'", "/Users/buh/WORK/logstash-plugins/logstash-input-twitter/lib/logstash/inputs/twitter/patches.rb:54:in `request'", "/Users/buh/tmp/git/logstash/vendor/bundle/jruby/1.9/gems/twitter-5.15.0/lib/twitter/streaming/client.rb:37:in `filter'", "/Users/buh/WORK/logstash-plugins/logstash-input-twitter/lib/logstash/inputs/twitter.rb:128:in `run'", "/Users/buh/tmp/git/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.0.0-java/lib/logstash/pipeline.rb:180:in `inputworker'", "/Users/buh/tmp/git/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.0.0-java/lib/logstash/pipeline.rb:174:in `start_input'"], :options=>nil, :level=>:warn}
```

And squid is showing a lot of this:

```
{"epochms":1447808629103,"response_time":0,"requesting_ip":"172.19.73.122","request_status":"TAG_NONE","next_hop_status":"400","reply_size":4296,"request_method":"NONE","request_url":"error:invalid-request","remote_ip":"-","mime_type":"text/html"}
```

The config used looks like this:

```
input {
  twitter {
    consumer_key => "REDACTED"
    consumer_secret => "REDACTED"
    oauth_token => "REDACTED"
    oauth_token_secret => "REDACTED"
    keywords => ["elasticsearch", "logstash", "kibana"]
    use_proxy => true
    proxy_address => "172.19.73.3"
    proxy_port => 3128
  }
}

filter {}

output {
  stdout { codec => rubydebug }
}
```

**Comment by purbon** [comment]
While trying to setup twitter gem playing plain non ssl with squid I end up having this problem:

```
The following error was encountered while trying to retrieve the URL: https://stream.twitter.com/1.1/statuses/filter.jsonUnsupported Request Method and Protocol
Squid does not support all request methods for all access protocols. For example, you can not POST a Gopher request.
```

it looks to me like an issue setting up squid, do you have any idea what might be? @untergeek @markwalkom

**Comment by purbon** [comment]
@markwalkom @untergeek got it working, just need a few tweeks and make squid compiled with ```--enable-ssl``` to get plenty of responses. yay!

**Comment by untergeek** [comment]
woot!  I'll test it soon.

**Comment by purbon** [comment]
@untergeek updating the PR now, will let you know.

**Comment by purbon** [comment]
With the changes introduced in https://github.com/purbon/logstash-input-twitter/commit/8c03cc67da6035c2d46fa65673beb14ae72f4ca5 , the twitter gem now does not force us to use an SSL connection when using a proxy. I've tested this with an small VM (ubuntu trusty) and squid3 and worked good for me.

Very important thing, is you are testing this with an ubuntu system, the available squid version does not support ssl, you should compile your own squid or use a third party PPA.

This is the sample squid3 config  I used:

```
http_port 3128

refresh_pattern ^ftp:           1440    20%     10080
refresh_pattern ^gopher:        1440    0%      1440
refresh_pattern -i (/cgi-bin/|\?) 0     0%      0
refresh_pattern .               0       20%     4320

acl localnet src 10.0.0.0/8     # RFC 1918 possible internal network
acl localnet src 172.16.0.0/12  # RFC 1918 possible internal network
acl localnet src 192.168.0.0/16 # RFC 1918 possible internal network
acl localnet src fc00::/7       # RFC 4193 local private network range
acl localnet src fe80::/10      # RFC 4291 link-local (directly plugged) machines

acl SSL_ports port 443          # https

acl Safe_ports port 80          # http
acl Safe_ports port 21          # ftp
acl Safe_ports port 443         # https
acl Safe_ports port 70          # gopher
acl Safe_ports port 210         # wais
acl Safe_ports port 1025-65535  # unregistered ports
acl Safe_ports port 280         # http-mgmt
acl Safe_ports port 488         # gss-http
acl Safe_ports port 591         # filemaker
acl Safe_ports port 777         # multiling http

acl CONNECT method CONNECT
acl POST method POST

http_access deny !Safe_ports
http_access deny CONNECT !SSL_ports
http_access allow localhost manager
http_access deny manager
http_access allow localnet
http_access allow localhost
http_access deny all
```

Example lines fetched with this changes:

```
I would like to talk to you about Islam…
https://t.co/E0n6uIdwf2

#BrusselsLockdown
#AMAs
#Politique
#QPPRRennes https://t.co/eFVZNi3Uqm
RT @carlosgomezgil: Cachonda la policía belga: saca esta imagen con comida para los gatos que ayer salieron en muchos tuits. https://t.co/E…
RT @hochstenbach: No #BrusselsLockdown without #LolCats https://t.co/E068Oc57Jg
claustrophobia is fun https://t.co/7MBQvTydJT https://t.co/XKwoNp5J1N
@Imaginesof5SOS https://t.co/uzfhXB8CRd unmute it, you won't be disappointed.
#AMAs
#BrusselsLockdown
#BlackFridayIn3Words
#SurvivorSeries
```

@untergeek @markwalkom could you please validate this? thanks a lot.

**Comment by purbon** [comment]
pushed upstream a general version of some of this patch https://github.com/sferik/twitter/pull/746, as soon as merged, some of this could be removed.

**Comment by untergeek** [comment]
It works for me, now!

LGTM™

**Pull request merged**